### ISSUE TITLE
Pusher fail to reconnect properly when coming back in foreground

### ISSUE THREAD

**Issue by pjechris** [pusher/pusher-websocket-swift#364]
### Steps to reproduce

- Go in background
- Wait a few seconds
- Go back in your app

### Expected behavior

Pusher reconnect and subscribe to channels

### Actual behavior

Connection fail with message "Invalid signature: Expected HMAC SHA256 hex digest" and no channel is subscribed

We did some digging. It seem that when coming back in foreground Pusher lib has hard time connecting and make many back and forth between `disconnected` and `connected` in a very short time. It lead to two things:

1. A lot of requests are triggered on our auth endpoint. They don't always arrive in correct order. This lead in having the message "Invalid signature: Expected HMAC SHA256 hex digest" because request _x-1_ is processed by Pusher while the socket id already changed to _x_
2. Not **all** requests are sent. We found out that, from time to time, we get a request for socket _x_. But Pusher fail ("invalid signature") expecting a signature for socket _y_. However we never see that socket id in our logs.

The reconnection attempts:
<img width="663" alt="Screenshot 2021-09-21 at 16 17 37" src="https://user-images.githubusercontent.com/7334856/134191687-b2eca8e7-ec5e-4e5a-965a-929b1fdd48c8.png">

The logs on the auth endpoint:
<img width="452" alt="Screenshot 2021-09-21 at 16 41 40" src="https://user-images.githubusercontent.com/7334856/134192948-b52a890b-6424-4c47-ae58-5ad05b45a70e.png">

The failure on iOS (see that we never received a request for 327215)
<img width="762" alt="Screenshot 2021-09-21 at 17 03 34" src="https://user-images.githubusercontent.com/7334856/134196140-6906455f-fc64-4de6-ad72-915798c835c3.png">

**Comment by benjamin-tang-pusher** [comment]
Thanks for this, I will take a look.

**Comment by benjamin-tang-pusher** [comment]
I believe you're right, it's possible the reconnection is causing the invalid signature error as the client gets a new socket ID. In your second screenshot (with the HTTP Status and PARAMS.SOCKET_ID columns), would it be possible to add an extra column for the token that was generated for that request? Would you also be able to export those iOS logs and attach them here?

**Comment by pjechris** [comment]
Unfortunately no because the generated token is not logged

**Comment by benjamin-tang-pusher** [comment]
Would you be able to add logging the token on your auth endpoint, along with the related screenshots similar to your first post, and we will escalate this.

**Comment by benjamin-tang-pusher** [comment]
Would you also be able to send us the code snippets ([in a secret gist if you prefer])(https://docs.github.com/en/github/writing-on-github/editing-and-sharing-content-with-gists/creating-gists) of your usage of the SDK?

**Comment by benjamin-tang-pusher** [comment]
Hi, we've been doing some extensive testing and we can't reproduce the connection flapping. Could you send us a video to see if there's anything we missed in our reproduction? Could you also send us some of your implementation in a secret gist?

**Comment by benjamin-tang-pusher** [comment]
I will close this ticket for now due to lack of response, please reopen if you need to revisit.

**Comment by pjechris** [comment]
@benjamin-tang-pusher I did send you the source code by mail on October 8th. And also said that I would give you the logs as soon as I can (which for now I can't).

FWIW we did some tests with v8 and it seem we have way less issues with this prior build

**Comment by pjechris** [comment]
@benjamin-tang-pusher can you confirm you received the code by mail? 🙏

**Comment by benjamin-tang-pusher** [comment]
Hi Chris, thanks for the information. I've responded to your team on the support ticket. To summarise, we are still investigating your case, but in the meantime, if version 8 works better for you, please use that as a workaround in the meantime.

**Comment by nicolas-vivot** [comment]
Hi,

Any news about this issue ? It's quite a critical bug and i'm a bit surprise that there is no fix yet.

We recently released an iOS application using the Pusher channel SDK to connect to private channel because it was working really fine for our Web application, but lately since more and more clients start to use the mobile application we have more and more cases where some of our in-app feature do not work as expected as they are based on Pusher communication and this bug appear depending on the situation.

I noticed that the temporary solution for @pjechris was to move back on version 8 of the SDK, so here are my questions for you @benjamin-tang-pusher :

- any update on that issue with a potential release soon ?
- is the version 8 SDK the same in term of API, would it be easy to move back to that version, what would be the cons in term of features compared to version 10 for just simple private channel connection & notification exchange use case ?

**Comment by miquelbeltran** [comment]
As commented on https://github.com/pusher/pusher-websocket-swift/issues/369 my guess is that this issure is related to removing the Reachability dependency and this autoconnect code: https://github.com/pusher/pusher-websocket-swift/commit/02aeab65c0ac9fb81ad131370159a9af0c90e946#diff-53e65cfa055a12c2f09666ac710e9eafc9643f5fb418cbfb5e0bae4bbac10410L321-L324

**Comment by mirrorlink** [comment]
@nicolas-vivot I'm also surprised they are not fixed yet. Like its been broken for about 5 or 6 releases. I've tested all of them

I also made a workaround now, using 8.0 with a fixed issue that was duplicating users on iOS

```
pod 'PusherSwift', :git => 'https://github.com/dganzella/pusher-websocket-swift.git', :branch => '8.0.1'

(if you have a podspec)
s.dependency 'PusherSwift'
```

**Comment by benw-pusher** [comment]
I am checking internally on the progress on this issue. We have attempted replication and have been unable to replicate so far which is causing some delays. 

In terms of version 8 of the SDK - this is the same in terms of the API so you should be able to easily move back to that version.

**Comment by nicolas-vivot** [comment]
Thanks for your reply.

We are already switching on the v8 yes, and will wait for your feedback about the bug fix on the support ticket or here.

**Comment by leofirespotter** [comment]
Has there been any update on this issue?

**Comment by pjechris** [comment]
This is a joke. No news since 6 months. No news on my last comment. And the ticket get closed.

**Comment by benw-pusher** [comment]
Looks like a bot closed this one out, reopening this.

**Comment by benjamin-tang-pusher** [comment]
I just found out a way to partially reproduce the Invalid signature error: I had to put the app into the background, and then switch from Wifi to LTE, and back again to Wifi, then resume the app into the foreground. This results in a HMAC digest error which I think is the same as what you're getting:

```
[PUSHER DEBUG] Network connection became unviable
2022-06-16 10:59:59.927034-0400 channels_ios_demo[88777:5227278] [tcp] nw_proto_tcp_route_init [C22.1:3] no mtu received
[PUSHER DEBUG] Network connection became unviable
2022-06-16 10:59:59.927858-0400 channels_ios_demo[88777:5227278] [tcp] nw_proto_tcp_route_init [C21.1:3] no mtu received
[PUSHER DEBUG] Network connection became unviable
2022-06-16 10:59:59.928613-0400 channels_ios_demo[88777:5227278] [tcp] nw_proto_tcp_route_init [C23.1:3] no mtu received
2022-06-16 10:59:59.931227-0400 channels_ios_demo[88777:5227278] Connection 24: encountered error(1:53)
...
...
State changed from disconnected to reconnecting
[PUSHER DEBUG] Attempting to reconnect after waiting 9.0 seconds (attempt 4)
State changed from reconnecting to connecting
[PUSHER DEBUG] Network connection became viable
[PUSHER DEBUG] websocketDidReceiveMessage {"event":"pusher:connection_established","data":"{\"socket_id\":\"135891.39615815\",\"activity_timeout\":120}"}
[PUSHER DEBUG] Socket established with socket ID: 135891.39615815
State changed from connecting to connected
[PUSHER DEBUG] sendEvent {"event":"pusher:subscribe","data":{"channel":"private-my-channel","auth":"dc9b99559fxxxxxx:5536608f52875b0c446a0c0826abc6c8373e0e73a7a35cf77c152755232aedd1"}}
[PUSHER DEBUG] Network connection became viable
[PUSHER DEBUG] websocketDidReceiveMessage {"event":"pusher:connection_established","data":"{\"socket_id\":\"135930.39523068\",\"activity_timeout\":120}"}
[PUSHER DEBUG] Socket established with socket ID: 135930.39523068
[PUSHER DEBUG] websocketDidReceiveMessage {"event":"pusher:error","data":{"code":null,"message":"Invalid signature: Expected HMAC SHA256 hex digest of 135930.39523068:private-my-channel, but got 5536608f52875b0c446a0c0826abc6c8373e0e73a7a35cf77c152755232aedd1"}}
```
The library connects twice upon resume, and there's confusion between the client and server over which socket should be authenticated. We're investigating why our library is queuing up the reconnection/resubscription, and I will update when we discover the cause.

**Comment by benjamin-tang-pusher** [comment]
Just to let you know we've assigned another engineer to investigate this.

**Comment by benjamin-tang-pusher** [comment]
We have a potential fix for this, which is currently in the testing stage.

**Comment by pjechris** [comment]
Hey thanks for the fix @benjamin-tang-pusher. We'll try it whenever possible.

For the multiple reconnection I also got the same issue last week when upgrading from Starscream 3 to 4. As they are using native websockets in this latest version and do you I'm starting to think this one might be an Apple bug? 😬

**Comment by benjamin-tang-pusher** [comment]
If its happening on Starscream as well, then its likely to be an Apple issue. We haven't looked too hard into the websocket networking yet.

**Comment by pjechris** [comment]
Ok. For now because of other priorities and code freeze period until end of the year I won't be able to test before January. I'll give you more feedback at that time.

Thanks

**Comment by pjechris** [comment]
@benjamin-tang-pusher 

I did some tests. I still see some HMAC errors but far less.

I also still have issues with socket not reconnecting from time to time. It seems when going in background it sometimes try to reconnect (and fail). After that the socket stays not connected until we start it again manually.

I'm more concerned about the reconnection issue than the HMAC error which is not blocking.

(logs when going in background)
<img width="1429" alt="Screenshot 2023-01-13 at 09 55 43" src="https://user-images.githubusercontent.com/7334856/212279129-4923eab4-7ba9-4abc-8dc4-454b65bc2546.png">

**Comment by benjamin-tang-pusher** [comment]
How did you know it happened when going into background? Were you able to occasionally reproduce the non reconnection?

**Comment by petarmarijanovicfive** [comment]
We also have occasional issues with this. Are there any updates with fixing this?

**Comment by benjamin-tang-pusher** [comment]
@petarmarijanovicfive Until we can reproduce this, we won't be able schedule when we can permanently fix this. I have heard of one other case of this with a similar pattern of error messages, but there's was a long running connection when without going into the background, so the reports seem conflicting. From your logging, can you tell whether this is occurring when the resume from background?

**Comment by mudassar69** [comment]
Is there any update on this issue? Our app is experiencing a similar problem where there is a delay of 2-3 minutes when transitioning from foreground to background. We are using the Pusher WebSocket React Native library (https://github.com/pusher/pusher-websocket-react-native), which is built on top of this library. Have there been any updates or solutions for this issue?

**Comment by benjamin-tang-pusher** [comment]
@mudassar69 they're seeing a "Software caused connection abort" then they have to manually restart their app. Are you seeing "Software caused connection abort" but it still eventually connects?

**Comment by benjamin-tang-pusher** [comment]
@pjechris Can you upgrade to 10.1.4, it now reconnects after receiving all POSIXErrors (timed out, could not be completed, software caused abort) as well a memory leak fix.

@mudassar69 If your reconnection is consistently taking a long time, then please open a separate GH issue as that wouldn't be the same issue as this one (where the reconnection fails outright)

**Comment by benjamin-tang-pusher** [comment]
I will close this ticket as there hasn't been further feedback on this issue.

Please use the latest versions of our libraries if anyone experiences this issue again.

**Comment by yu840915** [comment]
Hi, we have got this issue recently in `10.1.4`. What we found from the logs:

1. Pusher SDK connected with socket ID `29588.9277450` and behaved well.
2. App went to background and came back after a while.
3. Pusher SDK reconnected with new socket ID `29536.9278217`.
4. App tried to subscribe a channel.
5. Pusher SDK instructed app to auth the channel via `Authorizer` API, with new socket ID `29536.9278217`.
6. Our endpoint authorized the channel, we set the auth info back to the Pusher SDK.
7. Pusher SDK complained the channel should be authorized with old socket ID `29588.9277450`.
```
... Invalid signature: Expected HMAC SHA256 hex digest of 29588.9277450:...
```

Strangely, some channels with new socket ID `29536.9278217` successfully subscribed around the same time.
We have checked App ID and App key.
